### PR TITLE
Build: Shadow x-pack:protocol into x-pack:plugin:core

### PR DESCRIPTION
**Pull request by nik9000**
This bundles the x-pack:protocol project into the x-pack:plugin:core
project because we'd like folks to consider it an implementation detail
of our build rather than a separate artifact to be managed and depended
on. It is now bundled into both x-pack:plugin:core and
client:rest-high-level. To make this work I had to fix a few things.

Firstly, I had to make PluginBuildPlugin work with the shadow plugin.
In that case we have to bundle only the `shadow` dependencies and the
shadow jar.

Secondly, every reference to x-pack:plugin:core has to use the `shadow`
configuration. Without that the reference is missing all of the
un-shadowed dependencies. I tried to make it so that applying the shadow
plugin automatically redefines the `default` configuration to mirror the
`shadow` configuration which would allow us to use bare project references
to the x-pack:plugin:core project but I couldn't make it work. It'd *look*
like it works but then fail for transitive dependencies anyway. I think
it is still a good thing to do but I don't have the willpower to do it
now.

Finally, I had to fix an issue where Eclipse and IntelliJ didn't properly
reference shadowed transitive dependencies. Neither IDE supports shadowing
natively so they have to reference the shadowed projects. We fix this by
detecting `shadow` dependencies when in "Intellij mode" or "Eclipse mode"
and adding `runtime` dependencies to the same target. This convinces
IntelliJ and Eclipse to play nice.

### CODE REVIEW

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: plug -> project?

**Review by rjernst — APPROVED** [review]
This looks fine. A couple general comments:
* Can you add docs somewhere in contributing or something like that on the ability to shade and using the shadow vs compile?
* The name "shadow" for the configuration of what deps are *not* shaded. Is there a way to change this?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy
Comment (nik9000) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: ++

File: build.gradle
Comment (nik9000) [reply build.gradle]: Will do. It is ironic because and eclipse is caused by a shadow.

**Comment by nik9000** [comment]
> The name "shadow" for the configuration of what deps are not shaded. Is there a way to change this?

We'd have to make a new configuration. I was weary to do that at first but I think I understand how all things hangs together well enough to do that. What do you think of the name `unshaded`? What do you think of doing it in a follow up PR?

> Can you add docs somewhere in contributing or something like that on the ability to shade and using the shadow vs compile?

I'll add that before merging.

**Comment by nik9000** [comment]
And thanks for reviewing @rjernst!

**Review by alpar-t — COMMENTED** [review]
What is the advantage of using the shadow plugin as opposed to configuring the x-pack:core, high level rest and possibly other builds to simply include the class and resource files from the protocol project as in ` jar { from project(':x-pack:protocol').classes ; from project(':x-pack:protocol').processResources }` ( or something like that )  with a compile only dependency on `project(':x-pack:protocol')` to keep the IDEs happy about it could do this without leaking shadow plugin semantics across the build. It feels to me that we are throwing out too much of what the plugin does to warrant using it.  
We could even generalize that by creating a configuration of our own `bundled` or `baked` that can have only project level dependencies that behave like mentioned above: added to the jar and as compileOnly deps. Would probably keep things simpler in the gradle side.

**Comment by nik9000** [comment]
> What is the advantage of using the shadow plugin as opposed to configuring the x-pack:core, high level rest and possibly other builds to simply include the class and resource files from the protocol project as in jar { from project(':x-pack:protocol').classes ; from project(':x-pack:protocol').processResources } ( or something like that ) with a compile only dependency on project(':x-pack:protocol') to keep the IDEs happy about it could do this without leaking shadow plugin semantics across the build. It feels to me that we are throwing out too much of what the plugin does to warrant using it.
> We could even generalize that by creating a configuration of our own bundled or baked that can have only project level dependencies that behave like mentioned above: added to the jar and as compileOnly deps. Would probably keep things simpler in the gradle side.

I don't think it is worth re-implementing a bunch of the semantics of the shadow plugin though. We actually activate the META-INF merging across the board which is quite useful. And the jdbc driver uses relocation. So I think the shadow plugin is the right choice for us.

I don't think we can get away without leaking shadow semantics out of the build because we'd end up including the bundled classes twice in dependencies which will cause jar hell to fail. I'd *love* to alter the `default` configuration of a project when it has the shadow plugin so that we can just depend on the project and get the appropriate target. I just couldn't get it to work and got tired of fighting it.

I'll certainly look at your other points!

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy
Comment (nik9000) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy]: It mimics what we do above for the `runtime` configuration. I'm not arguing that it is right, just that it is in good company.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy
Comment (nik9000) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: That works! Awesome!

File: CONTRIBUTING.md
Comment (nik9000) [reply CONTRIBUTING.md]: Linked!

**Comment by atorok** [comment]
LGTM  Would be great if we could change the default configuration to shadow  but we can do it in another PR as discussed.

**Review by alpar-t — APPROVED** [review]

**Comment by nik9000** [comment]
Thanks for reviewing @atorok and @rjernst! I have two follow up things to do after I finish backporting this:

1. Figure out if I can make the `default` context work extend from the `shadow` context so you don't need to say `context: 'shadow'` all over the build.
1. Figure out if I can rename the `shadow` context into something more obvious like `unbundled`.

**Comment by rjernst** [comment]
I like the idea of having a `bundled` configuration, but all of these ideas are things that we can iterate on. If this works, let's go with it for now.

**Comment by nik9000** [comment]
I opened #32409 to rework the configuration of the shadow plugin. I think it makes it much more obvious what is going on.

**Pull request merged**
